src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormatPartitionWriter.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.cassandra.db.SerializationHeader;
+import org.apache.cassandra.db.rows.Unfiltered;
+import org.apache.cassandra.io.sstable.format.SortedTablePartitionWriter;
+import org.apache.cassandra.io.sstable.format.Version;
+import org.apache.cassandra.io.sstable.format.bti.RowIndexReader.IndexInfo;
+import org.apache.cassandra.io.util.SequentialWriter;
+
+/**
+ * Partition writer used by {@link BtiTableWriter}.
+ * <p>
+ * Writes all passed data to the given SequentialWriter and if necessary builds a RowIndex by constructing an entry
+ * for each row within a partition that follows {@link org.apache.cassandra.config.Config#column_index_size} of written
+ * data.
+ */
+class BtiFormatPartitionWriter extends SortedTablePartitionWriter
+{
+    private final RowIndexWriter rowTrie;
+    private final int indexSize;
+    private int rowIndexCount;
+
+    BtiFormatPartitionWriter(SerializationHeader header,
+                             ClusteringComparator comparator,
+                             SequentialWriter dataWriter,
+                             SequentialWriter rowIndexWriter,
+                             Version version)
+    {
+        this(header, comparator, dataWriter, rowIndexWriter, version, DatabaseDescriptor.getColumnIndexSize());
+    }
+
+
+    BtiFormatPartitionWriter(SerializationHeader header,
+                             ClusteringComparator comparator,
+                             SequentialWriter dataWriter,
+                             SequentialWriter rowIndexWriter,
+                             Version version,
+                             int indexSize)
+    {
+        super(header, dataWriter, version);
+        this.indexSize = indexSize;
+        this.rowTrie = new RowIndexWriter(comparator, rowIndexWriter);
+    }
+
+    @Override
+    public void reset()
+    {
+        super.reset();
+        rowTrie.reset();
+        rowIndexCount = 0;
+    }
+
+    @Override
+    public void addUnfiltered(Unfiltered unfiltered) throws IOException
+    {
+        super.addUnfiltered(unfiltered);
+
+        // if we hit the column index size that we have to index after, go ahead and index it.
+        if (currentPosition() - startPosition >= indexSize)
+            addIndexBlock();
+    }
+
+    @Override
+    public void close()
+    {
+        rowTrie.close();
+    }
+
+    public long finish() throws IOException
+    {
+        long endPosition = super.finish();
+
+        // the last row may have fallen on an index boundary already.  if not, index it explicitly.
+        if (rowIndexCount > 0 && firstClustering != null)
+            addIndexBlock();
+
+        if (rowIndexCount > 1)
+        {
+            return rowTrie.complete(endPosition);
+        }
+        else
+        {
+            // Otherwise we don't complete the trie. Even if we did write something (which shouldn't be the case as the
+            // first entry has an empty key and root isn't filled), that's not a problem.
```

Comment:
So the point here is that if we write exactly one block (or none), we don't need the trie, as we'd have to "slow scan" even the single block. Perhaps the comment could make that more clear, but that might just be me being dense :D